# Coding Exercises (Part 1)

Now, you will have the opportunity to analyze your own dataset. <br>
__Follow the instructions__ and insert your code! You are either requested to 
- Complete the Code and __Fill in the gaps__. Gaps are marked with "__---__" and are __placeholders__ for your code fragment. 
- Write Code completely __on your own__ 

In some exercises, you will find questions that can only be answered, if your code is correct and returns the right output! The correct answer is provided below your coding cell. There you can check whether your code is correct.

If you need a hint, check the __Hints Section__ at the end of this Notebook. Exercises and Hints are numerated accordingly.

If you need some further help or if you want to check your code, you can also watch the __solutions videos__ or check the __solutions notebook__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## DataFrame Basics III

### Exercise 8: apply(), MultiIndex and Strings

In [1]:
# run the cell!
import pandas as pd

In [2]:
# run the cell!
cars = pd.read_csv("../../Data/cars.csv")

In [32]:
# run the cell!
cars.head()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0    3504          12.0   
1  15.0          8         350.0       165.0    3693          11.5   
2  18.0          8         318.0       150.0    3436          11.0   
3  16.0          8         304.0       150.0    3433          12.0   
4  17.0          8         302.0       140.0    3449          10.5   

   model_year origin                       name  
0          70    usa  chevrolet chevelle malibu  
1          70    usa          buick skylark 320  
2          70    usa         plymouth satellite  
3          70    usa              amc rebel sst  
4          70    usa                ford torino

61. __Calculate__ for all __numerical columns__ the __range between highest and lowest__ value! __Fill in the gaps!__ The range for __mpg__ is...?

In [74]:
cars.iloc[:, :-2].apply(lambda x : x.max() - x.min(), axis=0)

TypeError: unsupported operand type(s) for -: 'str' and 'str'

The range for mpg is ... 37.6!

62. Create a __MultiIndex__ for the __cars__ DataFrame with the columns __"model_year"__ (outer level) and __"origin"__ (inner level). __Save__ the change!

In [34]:
cars.set_index(['model_year', 'origin'], inplace=True)

63. __Sort__ both index levels in __ascending__ order!. __Save__ the change!

In [35]:
cars.sort_index(ascending=True, inplace=True)

__Inspect__ Cars! What is the __name__ of the __second__ car?

In [36]:
# run the cell!
cars.head()

mpg  cylinders  displacement  horsepower  weight  \
model_year origin                                                      
70         europe  26.0          4          97.0        46.0    1835   
           europe  25.0          4         110.0        87.0    2672   
           europe  24.0          4         107.0        90.0    2430   
           europe  25.0          4         104.0        95.0    2375   
           europe  26.0          4         121.0       113.0    2234   

                   acceleration                  name  
model_year origin                                      
70         europe          20.5  vw 1131 deluxe sedan  
           europe          17.5           peugeot 504  
           europe          14.5           audi 100 ls  
           europe          17.5              saab 99e  
           europe          12.5              bmw 2002

The name of the second car is ... peugeot 504!

64. __Select__ all cars with the "model_year" __75__!

In [37]:
cars.loc[75]

mpg  cylinders  displacement  horsepower  weight  acceleration  \
origin                                                                    
europe  25.0          4          90.0        71.0    2223          16.5   
europe  29.0          4          90.0        70.0    1937          14.0   
europe  23.0          4         115.0        95.0    2694          15.0   
europe  23.0          4         120.0        88.0    2957          17.0   
europe  22.0          4         121.0        98.0    2945          14.5   
europe  25.0          4         121.0       115.0    2671          13.5   
japan   29.0          4          97.0        75.0    2171          16.0   
japan   24.0          4         134.0        96.0    2702          13.5   
japan   24.0          4         119.0        97.0    2545          17.0   
japan   33.0          4          91.0        53.0    1795          17.5   
usa     19.0          6         225.0        95.0    3264          16.0   
usa     18.0          6         250.0       105.0    3459          16.0   
usa     15.0          6         250.0        72.0    3432          21.0   
usa     15.0          6         250.0        72.0    3158          19.5   
usa     16.0          8         400.0       170.0    4668          11.5   
usa     15.0          8         350.0       145.0    4440          14.0   
usa     16.0          8         318.0       150.0    4498          14.5   
usa     14.0          8         351.0       148.0    4657          13.5   
usa     17.0          6         231.0       110.0    3907          21.0   
usa     16.0          6         250.0       105.0    3897          18.5   
usa     15.0          6         258.0       110.0    3730          19.0   
usa     18.0          6         225.0        95.0    3785          19.0   
usa     21.0          6         231.0       110.0    3039          15.0   
usa     20.0          8         262.0       110.0    3221          13.5   
usa     13.0          8         302.0       129.0    3169          12.0   
usa     23.0          4         140.0        83.0    2639          17.0   
usa     20.0          6         232.0       100.0    2914          16.0   
usa     23.0          4         140.0        78.0    2592          18.5   
usa     18.0          6         171.0        97.0    2984          14.5   
usa     19.0          6         232.0        90.0    3211          17.0   

                             name  
origin                             
europe                  vw dasher  
europe                  vw rabbit  
europe                 audi 100ls  
europe                peugeot 504  
europe                volvo 244dl  
europe                  saab 99le  
japan              toyota corolla  
japan               toyota corona  
japan                  datsun 710  
japan            honda civic cvcc  
usa       plymouth valiant custom  
usa                chevrolet nova  
usa               mercury monarch  
usa                 ford maverick  
usa              pontiac catalina  
usa             chevrolet bel air  
usa           plymouth grand fury  
usa                      ford ltd  
usa                 buick century  
usa     chevrolet chevelle malibu  
usa                   amc matador  
usa                 plymouth fury  
usa                 buick skyhawk  
usa           chevrolet monza 2+2  
usa               ford mustang ii  
usa                    ford pinto  
usa                   amc gremlin  
usa                 pontiac astro  
usa                    ford pinto  
usa                     amc pacer

65. __Select__ all cars with the "model_year" __76__ from __japan__! __How many__ cars do we get?

In [38]:
cars.loc[(76, 'japan')]

mpg  cylinders  displacement  horsepower  weight  \
model_year origin                                                      
76         japan   33.0          4          91.0        53.0    1795   
           japan   32.0          4          85.0        70.0    1990   
           japan   28.0          4          97.0        75.0    2155   
           japan   19.0          6         156.0       108.0    2930   

                   acceleration            name  
model_year origin                                
76         japan           17.4     honda civic  
           japan           17.0    datsun b-210  
           japan           16.4  toyota corolla  
           japan           15.5  toyota mark ii

There are ... 4 cars!

66. __Select__ all cars with the "model_year" __77__ and __79__  from __japan__! And only the columns __"mpg"__ and __"name"__! What is the __mpg__ of __datsun 210__?

In [39]:
cars.loc[[77,79], ['mpg', 'name']]

mpg                                  name
model_year origin                                            
77         europe  36.0                         renault 5 gtl
           europe  29.0                      vw rabbit custom
           europe  30.5                             vw dasher
           europe  21.5                              bmw 320i
           japan   31.5                     honda accord cvcc
           japan   33.5                 datsun f-10 hatchback
           japan   26.0               toyota corolla liftback
           japan   30.0                             subaru dl
           japan   22.0                            datsun 810
           japan   21.5                            mazda rx-4
           usa     30.0               buick opel isuzu deluxe
           usa     25.5                     plymouth arrow gs
           usa     17.5             chevrolet caprice classic
           usa     17.0            oldsmobile cutlass supreme
           usa     15.5                 dodge monaco brougham
           usa     15.0               mercury cougar brougham
           usa     17.5                    chevrolet concours
           usa     20.5                         buick skylark
           usa     19.0                plymouth volare custom
           usa     18.5                          ford granada
           usa     16.0                 pontiac grand prix lj
           usa     15.5          chevrolet monte carlo landau
           usa     15.5                      chrysler cordoba
           usa     16.0                      ford thunderbird
           usa     24.5                 pontiac sunbird coupe
           usa     25.5                   ford mustang ii 2+2
           usa     30.5                    chevrolet chevette
           usa     33.5                        dodge colt m/m
79         europe  31.9                      vw rabbit custom
           europe  25.4                    mercedes-benz 300d
           europe  27.2                           peugeot 504
           europe  37.3                    fiat strada custom
           japan   34.1                      mazda glc deluxe
           japan   31.8                            datsun 210
           usa     21.5                     pontiac lemans v6
           usa     19.8                      mercury zephyr 6
           usa     22.3                       ford fairmont 4
           usa     20.2                      amc concord dl 6
           usa     20.6                         dodge aspen 6
           usa     17.0             chevrolet caprice classic
           usa     17.6                       ford ltd landau
           usa     16.5                 mercury grand marquis
           usa     18.2                       dodge st. regis
           usa     16.9               buick estate wagon (sw)
           usa     15.5              ford country squire (sw)
           usa     19.2         chevrolet malibu classic (sw)
           usa     18.5  chrysler lebaron town @ country (sw)
           usa     35.7           dodge colt hatchback custom
           usa     27.4                         amc spirit dl
           usa     23.0                     cadillac eldorado
           usa     23.9     oldsmobile cutlass salon brougham
           usa     34.2                      plymouth horizon
           usa     34.5                  plymouth horizon tc3
           usa     28.4                 buick skylark limited
           usa     28.8                    chevrolet citation
           usa     26.8             oldsmobile omega brougham
           usa     33.5                       pontiac phoenix

The mpg of datsun 210 is ... 31.8!

67. __Swap the levels__ of the MultiIndex and then __sort__ the new MultiIndex in __ascending__ order! __Reassign cars__! __Fill in the gaps__!

In [76]:
cars = cars.swaplevel().sort_index(ascending=True)

TypeError: Can only swap levels on a hierarchical axis.

In [77]:
# run the cell!
cars

model_year  origin   mpg  cylinders  displacement  horsepower  weight  \
0            70  europe  26.0          4          97.0        46.0    1835   
1            70  europe  25.0          4         110.0        87.0    2672   
2            70  europe  24.0          4         107.0        90.0    2430   
3            70  europe  25.0          4         104.0        95.0    2375   
4            70  europe  26.0          4         121.0       113.0    2234   
..          ...     ...   ...        ...           ...         ...     ...   
393          82     usa  27.0          4         151.0        90.0    2950   
394          82     usa  27.0          4         140.0        86.0    2790   
395          82     usa  32.0          4         135.0        84.0    2295   
396          82     usa  28.0          4         120.0        79.0    2625   
397          82     usa  31.0          4         119.0        82.0    2720   

     acceleration                  name  
0            20.5  vw 1131 deluxe sedan  
1            17.5           peugeot 504  
2            14.5           audi 100 ls  
3            17.5              saab 99e  
4            12.5              bmw 2002  
..            ...                   ...  
393          17.3      chevrolet camaro  
394          15.6       ford mustang gl  
395          11.6         dodge rampage  
396          18.6           ford ranger  
397          19.4        chevrolet s-10  

[398 rows x 9 columns]

68. __Reset__ the __index__ of __cars__ and create a RangeIndex! __Save__ the change!

In [42]:
cars.reset_index(inplace=True)

In [78]:
# run the cell!
cars.head()

model_year  origin   mpg  cylinders  displacement  horsepower  weight  \
0          70  europe  26.0          4          97.0        46.0    1835   
1          70  europe  25.0          4         110.0        87.0    2672   
2          70  europe  24.0          4         107.0        90.0    2430   
3          70  europe  25.0          4         104.0        95.0    2375   
4          70  europe  26.0          4         121.0       113.0    2234   

   acceleration                  name  
0          20.5  vw 1131 deluxe sedan  
1          17.5           peugeot 504  
2          14.5           audi 100 ls  
3          17.5              saab 99e  
4          12.5              bmw 2002

69. __Convert__ all elements in the __origin__ column to __titlecase__! __Overwrite__ the __origin__ column!

# am  rescris coloana origin 

In [82]:
cars.origin = cars.origin.str.title()

In [83]:
# run the cell!
cars

model_year  origin   mpg  cylinders  displacement  horsepower  weight  \
0            70  Europe  26.0          4          97.0        46.0    1835   
1            70  Europe  25.0          4         110.0        87.0    2672   
2            70  Europe  24.0          4         107.0        90.0    2430   
3            70  Europe  25.0          4         104.0        95.0    2375   
4            70  Europe  26.0          4         121.0       113.0    2234   
..          ...     ...   ...        ...           ...         ...     ...   
393          82     Usa  27.0          4         151.0        90.0    2950   
394          82     Usa  27.0          4         140.0        86.0    2790   
395          82     Usa  32.0          4         135.0        84.0    2295   
396          82     Usa  28.0          4         120.0        79.0    2625   
397          82     Usa  31.0          4         119.0        82.0    2720   

     acceleration                  name  
0            20.5  vw 1131 deluxe sedan  
1            17.5           peugeot 504  
2            14.5           audi 100 ls  
3            17.5              saab 99e  
4            12.5              bmw 2002  
..            ...                   ...  
393          17.3      chevrolet camaro  
394          15.6       ford mustang gl  
395          11.6         dodge rampage  
396          18.6           ford ranger  
397          19.4        chevrolet s-10  

[398 rows x 9 columns]

Now, we want to __extract the manufacturer names__ from the column __"name"__. <br>
70. __Split__ the strings in the column __"name"__ on the very __first whitespace__ and __expand__ the splitted strings into __separate columns__! __Save__ new DataFrame in the variable __split__! __Fill in the gaps!__

In [62]:
split =  cars.name.str.split(" ", n=3, expand=True)

__Inspect__!

In [63]:
# run the cell!
split

0        1       2      3
0           vw     1131  deluxe  sedan
1      peugeot      504    None   None
2         audi      100      ls   None
3         saab      99e    None   None
4          bmw     2002    None   None
..         ...      ...     ...    ...
393  chevrolet   camaro    None   None
394       ford  mustang      gl   None
395      dodge  rampage    None   None
396       ford   ranger    None   None
397  chevrolet     s-10    None   None

[398 rows x 4 columns]

71. __Select__ the __first column__ and determine the __frequency of unique values__. __Fill in the gaps__! Which manufacturer is __most frequent__?

In [86]:
split[0].value_counts()

0
ford             51
chevrolet        47
plymouth         31
dodge            28
amc              28
toyota           26
datsun           23
vw               22
buick            17
pontiac          16
honda            13
mazda            12
mercury          12
oldsmobile       10
peugeot           8
fiat              8
audi              7
volvo             6
chrysler          6
renault           5
opel              4
saab              4
subaru            4
mercedes-benz     3
bmw               2
cadillac          2
triumph           1
ih                1
nissan            1
Name: count, dtype: int64

There are 51 cars from the manufacturer ... Ford!

72. Let´s __doublecheck__ the result from above. Check __how many__ car names in the column "names" __contain "ford"__!

In [87]:
cars.name.str.contains("ford").sum()

51

It´s ... again 51! It seems that the very first word in the column "name" is always the manufacturer name.

# Well Done!

------------------------------

# Hints (Spoiler!)

61. apply() method, maximum - minimum

62. set_index() method, pass list with column labels

63. sort_index() method

64. pass outer index labels to loc operator

65. cars.loc[(outer index labels, inner index labels)]

66. cars.loc[(outer index labels, inner index labels), column labels]

67. methods swaplevel() and sort_index()

68. reset_index() method

69. string(str)-method title()

70. string(str)-method split(), n = 1, 

71. value_counts() method 

72. string(str)-method contains()